clients/application-kafka/src/test/java/org/eclipse/hono/application/client/kafka/impl/KafkaApplicationClientImplTest.java

```diff
@@ -0,0 +1,232 @@
+    private MockConsumer<String, Buffer> mockConsumer;
+    private KafkaConsumer<String, Buffer> kafkaConsumer;
+    private KafkaConsumerConfigProperties consumerConfig;
+    private String tenantId;
+    private String deviceId;
+    private String correlationId;
+
+    static Stream<Type> messageTypes() {
+        return Stream.of(
+                Type.TELEMETRY,
+                Type.EVENT,
+                Type.COMMAND_RESPONSE);
+    }
+
+    /**
+     *
+     * Sets up fixture.
+     *
+     * @param vertx The vert.x instance to use.
+     */
+    @BeforeEach
+    void setUp(final Vertx vertx) {
+        final KafkaProducerConfigProperties producerConfig = new KafkaProducerConfigProperties();
+        final MockProducer<String, Buffer> mockProducer = KafkaClientUnitTestHelper.newMockProducer(true);
+        final CachingKafkaProducerFactory<String, Buffer> producerFactory = KafkaClientUnitTestHelper
+                .newProducerFactory(mockProducer);
+
+        tenantId = UUID.randomUUID().toString();
+        deviceId = UUID.randomUUID().toString();
+        correlationId = UUID.randomUUID().toString();
+
+        mockConsumer = new MockConsumer<>(OffsetResetStrategy.EARLIEST);
+        kafkaConsumer = KafkaConsumer.create(vertx, mockConsumer);
+
+        consumerConfig = new KafkaConsumerConfigProperties();
+        consumerConfig.setConsumerConfig(Map.of("client.id", "application-test-consumer"));
+        producerConfig.setCommonClientConfig(Map.of("client.id", "application-test-sender"));
+
+        client = new KafkaApplicationClientImpl(vertx, consumerConfig, producerFactory, producerConfig) {
+
+            //Override to use the mocked Kafka consumer
+            @Override
+            Future<MessageConsumer> createKafkaBasedDownstreamMessageConsumer(final String tenantId, final Type type,
+                    final Handler<DownstreamMessage<KafkaMessageContext>> messageHandler,
+                    final Handler<Throwable> closeHandler) {
+                return KafkaBasedDownstreamMessageConsumer
+                        .create(tenantId, type, kafkaConsumer, consumerConfig, messageHandler, closeHandler);
+            }
+        };
+    }
+
+    /**
+     * Cleans up fixture.
+     *
+     * @param context The vert.x test context.
+     */
+    @AfterEach
+    void shutDown(final VertxTestContext context) {
+        client.stop().onComplete(r -> context.completeNow());
+    }
+
+    /**
+     * Verifies that the message consumer created by the application client receives and processes the
+     * downstream message using the given message handler.
+     *
+     * @param msgType The message type (telemetry, event or command_response)
+     * @param ctx The vert.x test context.
+     */
+    @SuppressWarnings("unchecked")
+    @ParameterizedTest(name = PARAMETERIZED_TEST_NAME_PATTERN)
+    @MethodSource("messageTypes")
+    public void testCreateConsumerAcceptsMessage(final Type msgType, final VertxTestContext ctx) {
```

Comment:
I wonder if the test cases in this class are in the right place. It looks like we are actually testing the correct behavior of KafkaBasedDownstreamMessageConsumer or am I mistaken?

In this class, I would expect to find tests for creating a consumer and sending commands etc.

Reply:
In this PR for command response consumer, I have focused only on the consumers. In the above test, create consumer is also taken care of(eg. error in creating consumer leads to test failure). I think I went ahead and also tested if the messages are received by those consumers. If you think that this part is not needed here, I can remove it.

The next PR is for sending one way commands and then for synchronous command sending. In the consecutive PRs, I will add tests for sending commands.

Reply:
> If you think that this part is not needed here, I can remove it.

I do. As I said, I would expect this to be tested in *KafkaBasedDownstreamMessageConsumerTest* ...